### ISSUE TITLE
Laravel Vapor - Do you plan to support it?

### ISSUE THREAD

**Issue by jtomek** [statamic/ideas#312]
As you you have fully adopted Laravel (thanks, so happy you did that!), I would like to open an "Issue" and ask for your view on the Statamic support of Laravel Vapor. Laravel Vapor has become a flagship deployment system for the framework and many developers have adopted it. Unfortunately, it's currently not working with Statamic due to the filesystem design of AWS Lambda.

We managed to deploy Statamic to Vapor and get the website running. What we didn't figure out is how to facilitate the management experience Statamic is amazing at.

We had actually purchased Statamic 3(2) while it was in beta hoping this would get sorted out so we could easily rework our own website with business logic relating to our SaaS platform. But maybe we should look into buying Forge for Statamic? Although the whole server management (even with Forge) is something we would really like to avoid.

What's your view on this topic?

We love Statamic and hope we will be able to use it with Laravel Vapor which we love too. We actually use to run everything, including small helper functions, websockets or large ML models loaded from AWS Elastic FileSystem (EFS). If that's possible, there must be a way to make these two pieces of amazing tech work together ... maybe on the EFS? ..... :-)

Looking forward to hearing from you.

Edit: Grammar.

**Comment by JackEllis** [comment]
@jtomek You can attach EFS to Vapor. I have a video [in my course](https://serverlesslaravelcourse.com/) and [they document it here](https://docs.vapor.build/1.0/resources/storage.html#mounting-a-persistent-file-system).

I keep telling @jackmcdade that I'll record a video on it some day. The only problem I envision is running git to keep content in sync.

**Comment by jtomek** [comment]
@JackEllis Thanks for your reply...

I just attached a new EFS to my application but nothing changed. In the Vapor Docs, it says 

> Once mounted, you can store and retrieve files to the /mnt/local disk path. This path will be shared and accessible by all three of the Lambda functions
Source: https://docs.vapor.build/1.0/resources/storage.html#mounting-a-persistent-file-system

Does it mean that Statamic would have to use /mnt/local to manage its files? I added a new default local filesystem driver with root='/mnt/local', but there are so many methods inside statamic that it feels a bit unstable.

P.S. Are you saying if I get your course, I will find the answer to my problem there? Git would be of course nice but I can imagine that sync could be written as a Github Actions + Lambda function.

**Comment by JackEllis** [comment]
@jtomek Yeah the course shows you how to attach a filesystem. But you'd need to set-up a new filesystem driver and configure Statamic to use that driver. I don't have anything Statamic specific in my course and haven't been able to try it yet.

The root of the filesystem shouldn't affect Statamic, as they use Laravel's filesystem driver.

**Comment by mbryne** [comment]
I'd also love to register my interest in an official solution for Statamic and Vapor. We have gone all in on the platform and have had great success in doing so. 

We are currently using Nova for managing overall admin duties and website content across our sites but it's not a great editing experience for CMS style content. I love what we have seen of Statamic so far and would love to roll it out across all our sites.

I noticed from the nifty quote on the Serverless course that @JackEllis and @jackmcdade seem uniquely placed to offer some insight here as both the author of the Serverless course and Statamic. Is there any chance of fleshing out that video or getting some Vapor-specific Statamic document in relation to deploying on Vapor with an EFS filesystem? 

Is it a safe assumption that the need for the EFS filesystem etc may all be sidestepped by the Eloquent driver?

**Comment by anthonysexton** [comment]
Interested to see this (officially) supported also. @jackmcdade is this something that could be added as a sponsors goal?

**Comment by duncanmcclean** [comment]
Has anyone managed to get any further on this? We're considering using Vapor for one of our projects (which Statamic is hosting the marketing site of).

**Comment by jtomek** [comment]
We've been running it in production (midesk.co) on Vapor but are unhappy with how far we've gotten. 


If you have a chance to deploy it somewhere else, I would strongly recommend not investing your time into this topic unless it gets at least some attention from the Statamic team. It's been open since December 2019 and I haven't seen much interest.

If you want to go ahead with Vapor and Statamic anyway, our solution is underlined by the findings I have documented here:

https://statamic.com/forum/4379-vapor-laravel

**Comment by mbryne** [comment]
We haven't had a chance to deploy our Statamic site on Vapor yet but I just wanted to share an update on our current thought process and options.

Just a heads up that this is all based on our own requirements and skillsets. Obligatory your mileage may vary.

**~~EFS:~~**  

Seeing @jtomek EFS woes and not wanting to change too much of the default Vapor environment config we ruled this option out.

EFS = ❌

**~~Eloquent Driver~~**

We were waiting to see how the [Eloquent Driver](https://github.com/statamic/eloquent-driver) panned out but this currently only supports Entries. We would need Taxonomies at least to support dynamic creation of tags on the frontend, I started on a pull request but wasn't across Taxonomies enough and didn't have the time to wrap my head around it.

Eloquent driver = ❌

**~~Headless CMS / Gatsby~~**

Another potential option was to roll the site out on Gatsby and use Statamic as the backend CMS. The Gatsby learning curve and requirement of still hosting the CMS somewhere meant we just went for our last option.

Headless CMS / Gatsby = ❌

**GIT Integration / CI Server**

The last option we are going to pursue on this before direct support is offered is:

- **EC2 t3.nano instance**- Hosting instance of cms.oursite.com where content is managed. Located within Vapor VPC to access database when required.
- **[Statamic GIT Integration](https://statamic.dev/git-integration)** - GIT integration for committing changes made to content on that EC2 instance.
- **Github Actions / AWS CodePipeline** - Deploys to Vapor using the [Deploying from CI](https://docs.vapor.build/1.0/projects/deployments.html#deploying-from-ci) functionality.

We are going for this option as we need to automate those deployments anyway to sidestep local development / deployment issues.

GIT Integration / CI Server = ❓

---

So fingers crossed, I'll be working on that GIT / CI process sometime in the next 1-2 months and will document things as I go along.

**Comment by duncanmcclean** [comment]
Okay cool - we'll probably just opt for our normal app hosting setup (DO droplet & Forge). Hopefully it'll be easier to do in the future. 😄

**Comment by duncanmcclean** [comment]
@mbryne How did you get on with the Git integration / CI server setup - have you found that to work well? Considering the same thing.

**Comment by mbryne** [comment]
Hello @duncanmcclean I haven't had a chance as of yet but will be actioning that in the next 2-4 weeks. I am hoping it will be fairly straightforward. 

My thinking is to setup an EC2 instance to host the Statamic and backend Nova install within the existing Vapor network. I'll configure that to access our main Vapor serverless database directly where appropriate for Nova admin etc. I'll then use Statamic GIT commit functionality for content management.

I'm using both because there are things that Nova does well and things that Statamic does well. 

I'll deploy directly from GIT to Vapor using GIT Actions: https://github.com/marketplace/actions/laravel-vapor etc

I'll post back as I progress.

**Comment by mbryne** [comment]
Heads up @duncanmcclean you're not going to believe this, after considerable thought, effort and energy we are considering going back to WordPress as a headless CMS 🤯

Our thinking is:

- We are using blade templates heavily and have a modular content approach, no antlers etc, which means lots of data wrangling, even with [this](https://statamic.com/addons/silentz/blade) .
- We want that hybrid approach, most of our frontend functionality (user submitted content, forms, registrations, taxonomies etc) are already covered our Laravel installation and our database. 
- We want to use Statamic as our CMS only, with a slight integration between our taxonomies for content. 

I wish-o-wish Statamic went all in on that Eloquent Repository / Driver. We purchased our licence on the premise of that being developed but unfortunately it really was a sparse implementation that didn't work for our use case. Without that Eloquent driver we have been hampered by Vapor support etc. 

In short, the fact that it has been developed as a Laravel Package for V3 was A M A Z I N G but it strays far from sensible Laravel defaults and requires significant coupling with the rest of the codebase to get the most out of it. That leaves Statamic squarely in the 'Headless CMS' camp of which there are better suited options, for us personally. 

Statamic is a beautiful CMS, it wears its opinionated heart on its sleeve and its a fantastic product for its intended audience to be sure so no shade.  

Love your work!

**Comment by tomhatzer** [comment]
Hello everybody!

I've built a package that makes Statamic deployable and useable on Laravel Vapor.

You can find it here: https://github.com/tomhatzer/statamic-vapor-compatibility

Feedback is greatly appreciated, so if you have anything like ideas or if you find bugs, please share them there.

Thank you very much and have a great day.

**Comment by jezzdk** [comment]
Hey @tomhatzer, nice work on that package! I was wondering if you could add a little info to the readme? Specifically some background information about what problem it solves and how it solves it. I can read the code of course, but it would be great to hear your own thoughts about it :)

**Comment by tomhatzer** [comment]
Hey @jezzdk , yes sure. I'll update the readme and add a blog post soon. I'm currently working on a new website which will have topics about getting Laravel stuff running on Vapor. If you need any information in the meantime, send me a message on twitter. The link to my twitter handle is available in my github profile.
